src/createStore.js

```diff
@@ -163,38 +172,31 @@ export default function createStore(reducer, preloadedState, enhancer) {
    * return something else (for example, a Promise you can await).
    */
   function dispatch(action) {
-    if (!isPlainObject(action)) {
-      throw new Error(
-        'Actions must be plain objects. ' +
-        'Use custom middleware for async actions.'
-      )
-    }
-
-    if (typeof action.type === 'undefined') {
-      throw new Error(
-        'Actions may not have an undefined "type" property. ' +
-        'Have you misspelled a constant?'
-      )
-    }
-
-    if (isDispatching) {
-      throw new Error('Reducers may not dispatch actions.')
-    }
+    validateAction(action)
+    currentState = store.final.reducer(currentState, action)
+    store.final.onChange()
```

Comment:
I'm on the fence. I think, outside of the store, app code shouldn't depend on what actions were fired. But within a store enhancer or middleware, perhaps this isn't as true. Most middleware already is just  doing something based on what action was dispatched. Perhaps an enhancer overriding onChange is no different. The risk is if 2 or more enhancers override dispatch/onChange in incompatible ways, the more coupled the method signatures, the more likely of an incompatibility.